autopush/router/apns2.py

```diff
@@ -0,0 +1,144 @@
+        :param cert_file: Path to the PEM formatted APNs certification file.
+        :type cert_file: str
+        :param key_file: Path to the PEM formatted APNs key file.
+        :type key_file: str
+        :param topic: The *Bundle ID* that identifies the assoc. iOS app.
+        :type topic: str
+        :param alt: Use the alternate APNs publication port (if 443 is blocked)
+        :type alt: bool
+        :param use_sandbox: Use the development sandbox
+        :type use_sandbox: bool
+        :param max_connections: Max number of pooled connections to use
+        :type max_connections: int
+        :param logger: Status logger
+        :type logger: logger
+        :param metrics: Metric recorder
+        :type metrics: autopush.metrics.IMetric
+        """
+        self.server = SERVER if not use_sandbox else SANDBOX
+        self.port = 443 if not alt else 2197
+        self.ssl_context = hyper.tls.init_context(cert=(cert_file, key_file))
+        self.log = logger
+        self.metrics = metrics
+        self.topic = topic
+        self._max_connections = max_connections
+        self.connections = deque(maxlen=max_connections)
+        if self.log:
+            self.log.info("Starting APNS connection")
+
+        for i in xrange(0, max_connections):
+            self.connections.append(HTTP20Connection(
+                self.server, self.port,
+                ssl_context=self.ssl_context,
+                force_proto='h2'))
+
+    def send(self, router_token, payload, apns_id,
+             priority=True, topic=None, exp=None):
+        """Send the dict of values to the remote bridge
+
+        This sends the raw data to the remote bridge application using the
+        APNS2 HTTP2 API.
+
+        :param router_token: APNs provided hex token identifying recipient
+        :type router_token: str
+        :param payload: Data to send to recipient
+        :type payload: dict
+        :param priority: True is high priority, false is low priority
+        :type priority: bool
+        :param topic: BundleID for the recipient application (overides default)
+        :type topic: str
+        :param exp: Message expiration timestamp
+        :type exp: timestamp
+
+        """
+        body = json.dumps(payload)
+        priority = APNS_PRIORITY_IMMEDIATE if priority else APNS_PRIORITY_LOW
+        headers = {
+            'apns-id': apns_id,
+            'apns-priority': priority,
+            'apns-topic': topic or self.topic,
+        }
+        if exp:
+            headers['apns-expiration'] = exp
+        url = '/3/device/' + router_token
+        connection = self._get_connection()
+        try:
+            # request auto-opens closed connections.
+            stream_id = connection.request(
+                'POST', url=url, body=body, headers=headers)
+            response = connection.get_response(stream_id)
+            if response.status != 200:
+                reason = json.loads(response.read().decode('utf-8'))['reason']
```

Comment:
If not, then APNS is horribly broken and the JSON is probably corrupted as well. Might as well blow up early with something that'll get tossed into sentry.